examples/src/main/java/alluxio/cli/UnderFileSystemContractTest.java

```diff
@@ -0,0 +1,325 @@
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.cli;
+
+import alluxio.conf.InstancedConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.examples.S3ASpecificOperations;
+import alluxio.examples.UnderFileSystemCommonOperations;
+import alluxio.underfs.UfsFileStatus;
+import alluxio.underfs.UfsStatus;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.UnderFileSystemConfiguration;
+import alluxio.underfs.UnderFileSystemFactory;
+import alluxio.underfs.UnderFileSystemFactoryRegistry;
+import alluxio.underfs.options.DeleteOptions;
+import alluxio.util.ConfigurationUtils;
+import alluxio.util.io.PathUtils;
+
+import com.beust.jcommander.JCommander;
+import com.beust.jcommander.Parameter;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Integration tests for Alluxio under filesystems. It describes the contract of Alluxio
+ * with the UFS through the UFS interface.
+ */
+public final class UnderFileSystemContractTest {
+  private static final Logger LOG = LoggerFactory.getLogger(UnderFileSystemContractTest.class);
+  private static final String S3_IDENTIFIER = "s3";
+
+  @Parameter(names = {"--path"}, required = true,
+      description = "The under filesystem path to run tests against.")
+  private String mUfsPath;
+
+  @Parameter(names = {"--help"}, help = true)
+  private boolean mHelp = false;
+
+  private InstancedConfiguration mConf
+      = new InstancedConfiguration(ConfigurationUtils.defaults());
+
+  // The factory to check if the given ufs path is valid and create ufs
+  private UnderFileSystemFactory mFactory;
+  private UnderFileSystem mUfs;
+
+  private boolean mIsS3Ufs;
+
+  private UnderFileSystemContractTest() {}
+
+  private void run() throws Exception {
+    mFactory = UnderFileSystemFactoryRegistry.find(mUfsPath,
+        UnderFileSystemConfiguration.defaults(mConf));
+    // Check if the ufs path is valid
+    if (mFactory == null || !mFactory.supportsPath(mUfsPath)) {
+      LOG.error("{} is not a valid path", mUfsPath);
+      System.exit(1);
+    }
+
+    mConf.set(PropertyKey.UNDERFS_LISTING_LENGTH, "50");
+    mConf.set(PropertyKey.USER_BLOCK_SIZE_BYTES_DEFAULT, "512B");
+    // Increase the buffer time of journal writes to speed up tests
+    mConf.set(PropertyKey.MASTER_JOURNAL_FLUSH_BATCH_TIME_MS, "1sec");
+
+    createUnderFileSystem();
+    mIsS3Ufs = mUfs.getUnderFSType().equals(S3_IDENTIFIER);
```

Comment:
nit: i prefer not having this redundant field. is it expensive to query `getUnderFSType`?